identity-service/src/typed-routes/solana/layout-utils.ts

```diff
@@ -0,0 +1,92 @@
+import {
+  Blob,
+  Layout,
+  blob,
+  struct,
+  u32,
+  uint8ArrayToBuffer
+} from '@solana/buffer-layout'
+
+/**
+ * Wrapper that makes Ethereum wallet addresses encoded and decoded as 20 byte blobs
+ */
+export class EthereumAddress extends Layout<string> {
+  private blob: Blob
+  constructor(property?: string) {
+    super(20, property)
+    this.blob = blob(20, property)
+  }
+
+  /** @override */
+  getSpan(b: Uint8Array, offset?: number) {
+    return this.blob.getSpan(b, offset)
+  }
+
+  /**
+   * Proxies the decoding to the underlying Blob, then
+   * converts the buffer to hex and prepends '0x'.
+   * @override
+   * */
+  decode(b: Uint8Array, offset = 0): string {
+    const buffer = this.blob.decode(b, offset)
+    return '0x' + Buffer.from(buffer).toString('hex')
+  }
+
+  /**
+   * Strips the 0x and converts the address to a buffer, then
+   * proxies the encoding to the underlying Blob.
+   * @override
+   * */
+  encode(src: string, b: Uint8Array, offset: number): number {
+    const strippedEthAddress = src.replace('0x', '')
+    // Need to pad the array to length 20 - otherwise, hex eth keys starting with '0' would
+    // result in truncated arrays, while eth spec is always 20 bytes
+    const buf = Buffer.from(strippedEthAddress, 'hex')
+    const fixedBuf = Buffer.alloc(20, 0)
+    buf.copy(fixedBuf, 20 - buf.length)
+    return this.blob.encode(buf, b, offset)
+  }
+}
+
+/**
+ * Wrapper that encodes strings the way Borsh does, with the length prepended
+ */
+export class BorshString extends Layout<string> {
+  constructor(maxLength: number, property?: string) {
+    super(u32().span + maxLength, property)
+  }
+
+  getSpan(b: Uint8Array, offset = 0): number {
+    if (!b) {
+      return this.span
+    }
+    const length = u32().decode(b, offset)
+    return u32().span + length
+  }
+
+  decode(b: Uint8Array, offset = 0): string {
+    const length = u32().decode(b, offset)
+    const value = blob(length).decode(b, offset + u32().span)
+    return uint8ArrayToBuffer(value).toString('utf-8')
+  }
+
+  encode(src: string, b: Uint8Array, offset: number): number {
+    const srcb = Buffer.from(src, 'utf-8')
+    if (srcb.length > this.span) {
+      throw new RangeError('text exceeds maxLength')
+    }
+    if (offset + srcb.length > b.length) {
+      throw new RangeError('text length exceeds buffer')
+    }
+    return (
+      u32().encode(srcb.length, b, offset) +
+      blob(srcb.length).encode(srcb, b, offset + u32().span)
+    )
+  }
+}
```

Comment:
Lots of buffer manipulation and conversion in this file. How much of this is directly copied from known working code? I don't trust my ability to find any subtle bugs that might creep out of replacing this type of logic. 😅

Reply:
These do deserve some more thorough testing. I think they're ok for this PR because they're actually unused (just used to make mock instructions for tests), but before we adopt using them for actual instruction creation we should have the tests here to ensure accuracy.

That said, 90% of this code is just wrapping existing buffer layout things to handle getting the strings back for ease of use. I'm not doing much actual manipulation myself.

As an aside, this `BorshString` one is interesting, though - what happened here is that the existing encodings from buffer-layout (null terminated cstr and variable length utf8) are different from the way Borsh serialization for strings work. Borsh serializes the length of the string at the start as a u32 and then uses that for deserializing. Notably when using this class I opted to just construct it with a max size and then have the caller be in charge of trimming the buffers to the proper size so there's no trailing 0s.

Reply:
https://linear.app/audius/issue/PAY-1942/migrate-instruction-helpers-from-identity-to-their-own-package

Reply:
It's like you read my mind with the separate package idea ;)